presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java

```diff
@@ -1207,6 +1226,66 @@ public void testBadUrlMissingSchema()
         }
     }
 
+    @Test
+    public void testQueryCancellation()
+            throws Exception
+    {
+        CountDownLatch queryStarted = new CountDownLatch(1);
+        CountDownLatch queryFinished = new CountDownLatch(1);
+        AtomicReference<String> queryId = new AtomicReference<>();
+        AtomicReference<Throwable> queryFailure = new AtomicReference<>();
+
+        Future<Void> queryFuture = executorService.submit(() -> {
+            try (Connection connection = createConnection("blackhole", "default")) {
+                try (Statement statement = connection.createStatement()) {
+                    try (ResultSet resultSet = statement.executeQuery("SELECT * FROM test_cancellation")) {
+                        queryId.set(resultSet.unwrap(PrestoResultSet.class).getQueryId());
+                        queryStarted.countDown();
+                        try {
+                            resultSet.next();
+                        }
+                        catch (Throwable t) {
+                            queryFailure.set(t);
+                        }
+                        finally {
+                            queryFinished.countDown();
+                        }
+                    }
+                }
+            }
+            return null;
+        });
+
+        queryStarted.await(1, MINUTES);
+        assertNotNull(queryId.get());
+        assertQueryState(queryId.get(), "QUEUED", "PLANNING", "STARTING", "RUNNING");
+        queryFuture.cancel(true);
+        queryFinished.await(1, MINUTES);
+        assertNotNull(queryFailure.get());
+        assertQueryState(queryId.get(), "FAILED");
```

Comment:
Unfortunately no. When query is cancelled by user it will be in `FAILED` state. We either have to add specific state for canceled queries, or add the "failure_cause" column to the `queries` table.